Transpile to Kotlin: schedule package

isiko opened the pull request:
# Description
The `nerd.tuxmobil.fahrplan.congress.schedule` had still some .java files. As the Standard Language for Android Apps is now Kotlin, I transpiled them to Kotlin. I used the transpiling tool, provided by Android Studio and made some cleanups afterward. This also resulted in removing some code which was deprecated since cc9817e4710a50ab526c7f92f94ae7aa9d9c389e

# Fully tested on (with all Build flavors):
- Tablet (Pixel C), API lvl 24 Emulator
- Tablet (Pixel C), API lvl 30 Emulator
- Phone (Pixel 5), API lvl 31 Emulator
- Phone (Pixel 2), API lvl 24 Emulator

johnjohndoe commented:
:mag_right: Please rebase your branch onto the latest commit of `master` and review it.

1. `git checkout master` -> `git pull origin master` // `origin` being **this** repository
2. `git rebase transpilingSchedulePackage master`
3. See below
4. `git push --force origin master`

Please review your branch to see if something can be update after rebasing onto formerly merged branches. Also, check for the usually improvements such as nullability, return types, ... which I mentioned in other reviews.

johnjohndoe commented:
Please note that your `master` branch is **not** up-to-date. You need to add **this** repository as a second remote in order to update your fork - if you did not do so. Check yourself via `git remote -v`

1. `git remote add EventFahrplan https://github.com/EventFahrplan/EventFahrplan.git` // only needed do be done once
2. `git fetch EventFahrplan`
3. `git checkout master`
4. `git pull EventFahrplan master`

Now **your local** `master` branch is up-to-date.

5. `git push --force origin master` :warning: Here `origin` refers to **your** remote repository aka. `https://github.com/isiko/EventFahrplan`. Check yourself via `git remote -v`.

Now **your remote** `master` branch is up-to-date.

6. Now you can rebase your feature branch once again and push it to GitHub.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    fun scrollToColumn(col: Int, fast: Boolean) {
+        val max = if (this.columnWidth == 0) 0 else (getChildAt(0).measuredWidth - columnWidth) / columnWidth
```
`this.columnWidth` -> `columnWidth`

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,266 @@
+    }
+
+    fun getColumnWidth(): Int {
+        check(roomsCount != NOT_INITIALIZED) { "The \"roomsCount\" field must be initialized before invoking \"getColumnWidth\"." }
```
This becomes much nicer with a raw string:

``` kotlin
check(roomsCount != NOT_INITIALIZED) {
    """The "roomsCount" field must be initialized before invoking "getColumnWidth"."""
}
```

johnjohndoe reviewed (changes requested):
Thank you for the conversion. :+1: 

I pointed out a few things to improve. Please additionally try to clean up all **one letter** variable or parameter names even though they have been like this in the Java file.

johnjohndoe commented:
Since there a still a few `getChildAt(0)` calls you might want to introduce a local extension function such as:

``` kotlin
private val ViewGroup.firstChild
    get() = get(0) as ViewGroup
```

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,266 @@
+        val max = if (this.columnWidth == 0) 0 else (getChildAt(0).measuredWidth - columnWidth) / columnWidth
+        val col = min(max(col, 0), max)
+
+        val scrollTo = col * columnWidth
```
I would never shadow another variable on purpose. This could lead to unforseen error when refactoring. Why not `columnIndex`?

isiko replied:
Because the new Variable is just the old one but Clamped to a specific Range (between 0 and the calculated maximum Value). In Java, you could just change the value of the Parameter, but for some reason, Parameters are read-only in Kotlin. I'll change it to `columnIndex`

johnjohndoe replied:
In Java one would make parameters `final` to guarantee that the function does not modify the input which can be very dangerous. Thanks for the change.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,270 @@
+                        activeColumnIndex + columnDistance
+                    }
+                } else {
+                    if (abs(distance) > columnWidth / 4) {
```
Can we please move the magic number `4` into a constant? What does it mean? What does it stand for? The name of the constant should clarify this.

johnjohndoe replied:
Please use `const` and move it to the companion object and capitalize it, e.g. `MIN_PORTRAIT_SCROLLING_PERCENTAGE`. Why not `0.25` instead of `1 / 4`?

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,270 @@
+    var activeColumnIndex = 0
+        private set
+
+    private val ViewGroup.firstChild
```
You can move this outside of the class, to the bottom of the file.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,270 @@
+            }
+            val availableDips = availablePixels.toFloat() / densityScaleFactor
+            val minWidthColumnCount = floor((availableDips / minColumnWidthDip).toDouble())
+                .toInt()
```
This still fits in the line.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,270 @@
+        if (roomNames != null) {
+            val firstRoomNameView = roomNames!!.firstChild
+            firstRoomNameView.children.forEach {
+                it.updateLayoutParams { width = columnWidth }
```
Maybe `it.layoutParams.width = columnWidth` is good enough. We use it up there, too. Consistency rules.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,270 @@
+
+        container.measure(
+            MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED),
+            MeasureSpec.makeMeasureSpec(0, MeasureSpec.UNSPECIFIED)
```
Can this be de-duplicated?

isiko replied:
I wouldn't know how, as one defines the height and the other one the width and It seems like there is no version of the function that takes both Parameters as one. It would be possible to add the value as a constant and pass it twice, although that would probably be bad for readability while not really making meaningful improvements

johnjohndoe replied:
OK. Let's leave it as is.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,270 @@
+
+                var tempColumnIndex = activeColumnIndex
+
+
```
You added quite a lot of whitespace to the file. Let reduce it a bit all over.

johnjohndoe commented:
Please clean up the commit message (typo) and unneeded text.

isiko commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,270 @@
+            return super.onDown(event)
+        }
+
+        override fun onFling(
```
I read the Documentation, but couldn't find a catchy name 😅 My best solution would be `downMotionEvent` and `triggerMotionEvent`, but I'm not really happy with that either 😕

johnjohndoe replied:
It's okay to me to leave it as is.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,266 @@
+
+        val container = firstChild
+        container.children.forEach {
+            // Set width for each Child to columnWidth
```
The comment becomes very redundant now. Let's drop it.

johnjohndoe commented:
![image](https://user-images.githubusercontent.com/144518/160702130-362e6230-3524-4dc3-866c-3ac11cd07ebd.png)

Did you notice that you commit message has grown with every push? :arrow_up:

johnjohndoe commented:
![image](https://user-images.githubusercontent.com/144518/160702471-38844a63-3bc6-4083-bb76-e727d0c54ea9.png)

Did you see the latest conflicts? :arrow_up:

isiko commented:
> Did you notice that you commit message has grown with every push? ⬆️

Actually no, thanks for pointing that out. For some reason, the Github UI didn't show me that. Should be fixed in the latest commit,.

---

> Did you see the latest conflicts? ⬆️

Yes, but I was a bit too tired to get to that yesterday. I will try to resolve those when I get home 👍

johnjohndoe commented:
> Yes, but I was a bit too tired to get to that yesterday. I will try to resolve those when I get home +1

No worries. It is just the logging cleanup (see c2131f72af9f1e4bbef6b4d983291a1037012993) which conflicts with your changes. Should be easy to apply.

isiko commented:
Had some minor struggles with fixing the conflicts as it was a bit more complex than the conflicts I had in the past, but should be done now 👍

johnjohndoe reviewed (changes requested):
Please test the app on a device before you update/push the branch. The current version is broken. When I scroll horizontally then the room names get stuck together the more I scroll to the right.

Please also re-add the logging statements which were commented out in the Java file. They are there for easy debugging purpose and not dead code in this case.

isiko commented:
> Please test the app on a device before you update/push the branch. The current version is broken. When I scroll horizontally then the room names get stuck together the more I scroll to the right.

Sorry, should have tested that 😅 For me the error is fixed now, although I should have seen that earlier 🤔
(Tested with all the Devices listed)

> Please also re-add the logging statements which were commented out in the Java file. They are there for easy debugging purpose and not dead code in this case.

Those should also be back now, somehow they must have been removed while converting from Java to Kotlin

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,269 @@
+    private val logging: Logging = Logging.get()
+    private var xStart = 0
+    private var roomsCount = NOT_INITIALIZED
+    private var roomNames: HorizontalScrollView? = null
```
Did you try if `roomNames` could be non-nullable? It is set in `FahrplanFragment` via `snapScroller.setChildScroller(roomScroller)` where `roomScroller` is guaranteed to be non-null. Not sure if anyone else is accessing `roomNames` before.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,269 @@
+        )
+
+        if (roomNames != null) {
+            roomNames!!.firstChild.children.forEach {
```
Did I incorrectly convert this to Kotlin before causing the room names bug?

isiko replied:
That section caused the bug, but it wasn't a result of your Transpilation. I just made a mistake while copying it 🤦

isiko commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,269 @@
+    private var roomsCount = NOT_INITIALIZED
+    private var roomNames: HorizontalScrollView? = null
+    private var displayColumnCount: Int
+    private var columnWidth: Int
```
That's probably not possible, as that getter is only used outside the class, but there are some places inside the class where the attribute is used directly and which would break. ([here](https://github.com/EventFahrplan/EventFahrplan/blob/c3b6203746b901bfa52bea3d24a6b9c99827089d/app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.java#L121) for example)

johnjohndoe replied:
I think the internal and external read usage should not be different. Both rely on that the property is initialized.

You can pull up the initialization from the `init { }` block.

``` kotlin
var columnWidth: Int = 0
    get() {
        check(roomsCount != NOT_INITIALIZED) {
            """The "roomsCount" field must be initialized before invoking "getColumnWidth"."""
        }
        return field
    }
    private set
```

Not tested!

isiko replied:
I thought so too, but it seems like the `onSizeChanged` method is called before the `roomCount` is initialized [here](https://github.com/EventFahrplan/EventFahrplan/blob/c3b6203746b901bfa52bea3d24a6b9c99827089d/app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/FahrplanFragment.kt#L219). This should probably be looked into, but it seems that this is an issue that is a bit out of the scope of just transpiling.

johnjohndoe replied:
Interesting. Thank you for the clarification. I agree - this is out of scope.

johnjohndoe commented:
No need to rebase again. Once I assign myself I will take it from there. Too many pushes cost lot of credits on the CI pipeline..

isiko commented:
> No need to rebase again. Once I assign myself I will take it from there. Too many pushes cost lots of credits on the CI pipeline..

Sorry, didn't see that 😅

johnjohndoe reviewed (changes requested):
I found an error in the current version:

# Horizontal sync error
- The room names and the session become out of sync when I scroll to the right edge of the schedule. The screenshot shows how the columns do not align.
  ![image](https://user-images.githubusercontent.com/144518/161111502-3e73efcd-40af-4314-9cf2-692dc0e88526.png)

## Test setup
- `rc3` flavor, `debug` build
- Nexus 9 emulator (tablet), API 22

## Crosscheck
- I cannot reproduce the error on `master` (aa89076e69d77eb393d8979a9a9fd5f44f81ab2a).

isiko commented:
I found the Error and have the fix.

---

Also, I would suggest a minor improvement to the `scrollToColumn` Method; The Old Method of calculating the Maximum Column that can be scrolled to is less intuitive and also didn't calculate the right Value (in the Bug you found it should have always snaped back). Should I commit that? I'm asking because I do not want to make an unnecessary commit if you can spot a flaw with this change at first glance.

Old:
```val maxColValue = if (columnWidth == 0) 0 else (firstChild.measuredWidth - columnWidth) / columnWidth```

New:
```val maxColValue = if (columnWidth == 0) 0 else roomsCount - displayColumnCount```

johnjohndoe commented:
That was quick. Cool.

I won't be able to look into this until next week, though. So take you time to check yourself.

I suggest you add the fix to the existing commit and put the modification in a separate commit. Please state in the commit message why you change that particular part.

isiko commented:
I'm improving the `onFling` Method and was wandering what [this](https://github.com/EventFahrplan/EventFahrplan/blob/aa89076e69d77eb393d8979a9a9fd5f44f81ab2a/app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.java#L80) `try/catch`-Statement is trying to catch?
The only thing in there that could throw an Error is the `scrollToColumn` Call, but all other Calls of that Method do not have a try/catch. Looking at the git history it seems like this was added in eded8d6ade9b865137a9f422c5247fe4c388d63f but without any explanation for the `try/catch`.

johnjohndoe commented:
> I'm improving the `onFling` Method and was wandering what [this](https://github.com/EventFahrplan/EventFahrplan/blob/aa89076e69d77eb393d8979a9a9fd5f44f81ab2a/app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.java#L80) `try/catch`-Statement is trying to catch? [...]

I could neither find out about the motivation to wrap the statement in a try/catch block. I imagined one of the `Math` functions could through an `ArithmeticException` but the ones being used do not.

johnjohndoe commented on app/src/main/java/nerd/tuxmobil/fahrplan/congress/schedule/HorizontalSnapScrollView.kt:
```diff
@@ -0,0 +1,239 @@
+            velocityY: Float
+        ): Boolean {
+            val normalizedVelocityX = velocityX / resources.displayMetrics.density
+            val columns = ceil((normalizedVelocityX / SWIPE_THRESHOLD_VELOCITY * 3).toDouble()).toInt()
```
> `SWIPE_THRESHOLD_VELOCITY * 3`

The meaning of `3` should be clarified by declaring a constant with a descriptive name for it.

isiko replied:
should be fixed

johnjohndoe replied:
Do you know why the value is `3`?

isiko replied:
I have no idea. It seems to be some multiplication with which the amount of columns scrolled is multiplied to scroll further, hence the name, but I don't know why `3` was chosen.

johnjohndoe replied:
Maybe, `3` has been chosen as a value to **subjectively** satisfy **some amount of flinging** over rooms (when there is a lot) as discussed in #22.

isiko replied:
That's probably it 👍

johnjohndoe reviewed (commented):
I tried out the current state (Pixel 6, Android 12 and Tablet emulator, Android 5.1.1). I could not find any errors. :heart_eyes: Thanks for fixing the earlier issue.

The code on `master` and the code on this branch is almost **not unit tested**. Did you have a look if it is possible to cover it? Being backed by unit tests **before** would feel much better when changing these calculations.

I also would love to schedule a **review with you** to talk through some of the details on this branch. Let's arrange something via mail.

isiko commented:
> The code on `master` and the code on this branch is almost **not unit tested**. Did you have a look if it is possible to cover it? Being backed by unit tests **before** would feel much better when changing these calculations.

I'd love to, only problem is that I've never actually written unit tests for Android-Apps 😅 I could, of course, teach myself, but unfortunately I do not have time for that at the moment.

> I also would love to schedule a **review with you** to talk through some of the details on this branch. Let's arrange something via mail.

Great, I'll send you a mail 👍

isiko commented:
Implemented the discussed Changes:
- Improved commit messages
- Exported Velocity and Distance Checks to separate Methods
- renamed `maxColValue` ->`maxColumnIndex`
- Added explanation for `FLING_COLUMN_MULTIPLIER`

johnjohndoe commented:
@isiko I figured out a way to enable that the mathematical calculations which can be found in `HorizontalSnapScrollView` can be unit tested. :partying_face: Please have a look at the commits on the [HorizontalSnapScrollView](https://github.com/EventFahrplan/EventFahrplan/tree/HorizontalSnapScrollView) branch when you have time. I think it should be a good starting point to (1) add **unit tests**, then (2) redo the **Kotlin conversion** and after that (3) apply **further modifications**. We can do this in a pair-programming session if you like. As mentioned offline - no hurries!

I suggest leaving this branch / PR **as is** because it provides a **valuable state** to compare with. I would not create a new PR but just work on a new branch until we reach the same state as here. Once we are satisfied we could replace this branch and thereby update this PR.

isiko commented:
Looks nice, I like the Idea of splitting the Class. I'd love to work on that, allthough that will probably be in about 4 Weeks :+1:

isiko commented:
Ths somehow got closed while switching Branches :confused:
